server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebPushContract.scala

```diff
@@ -0,0 +1,508 @@
+      .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .build()
+  }
+
+  private def getPushServerUrl(pushServer: ClientAndServer): String =
+    s"http://127.0.0.1:${pushServer.getLocalPort}$PUSH_URL_PATH"
+
+  // return pushSubscriptionId
+  private def createPushSubscription(pushServer: ClientAndServer): String =
+    `given`
+      .body(
+        s"""{
+           |    "using": ["urn:ietf:params:jmap:core"],
+           |    "methodCalls": [
+           |      [
+           |        "PushSubscription/set",
+           |        {
+           |            "create": {
+           |                "4f29": {
+           |                  "deviceClientId": "a889-ffea-910",
+           |                  "url": "${getPushServerUrl(pushServer)}",
+           |                  "types": ["Mailbox"]
+           |                }
+           |              }
+           |        },
+           |        "c1"
+           |      ]
+           |    ]
+           |  }""".stripMargin)
+    .when
+      .post
+    .`then`
+      .statusCode(SC_OK)
+      .extract()
+      .jsonPath()
+      .get("methodResponses[0][1].created.4f29.id")
+
+  private def updateValidateVerificationCode(pushSubscriptionId: String, verificationCode: String): String =
+    `given`()
+      .body(
+        s"""{
+           |    "using": ["urn:ietf:params:jmap:core"],
+           |    "methodCalls": [
+           |      [
+           |        "PushSubscription/set",
+           |        {
+           |            "update": {
+           |                "$pushSubscriptionId": {
+           |                  "verificationCode": "$verificationCode"
+           |                }
+           |              }
+           |        },
+           |        "c1"
+           |      ]
+           |    ]
+           |  }""".stripMargin)
+    .when
+      .post
+    .`then`
+      .statusCode(SC_OK)
+      .contentType(JSON)
+      .extract()
+      .body()
+      .asString()
+
+  private def sendEmailToBob(server: GuiceJamesServer): Unit = {
+    val smtpMessageSender: SMTPMessageSender = new SMTPMessageSender(DOMAIN.asString())
+    smtpMessageSender.connect("127.0.0.1", server.getProbe(classOf[SmtpGuiceProbe]).getSmtpPort)
+      .authenticate(ANDRE.asString, ANDRE_PASSWORD)
+      .sendMessage(ANDRE.asString, BOB.asString())
+    smtpMessageSender.close()
+
+    awaitAtMostTenSeconds.until(() => server.getProbe(classOf[SpoolerProbe]).processingFinished())
+  }
+
+  private def setupPushServerCallback(pushServer: ClientAndServer, bodyRequestOnPushServer: AtomicReference[String]): Unit =
+    pushServer
+      .when(request
+        .withPath(PUSH_URL_PATH)
+        .withMethod("POST"))
+      .respond(new ExpectationResponseCallback() {
+        override def handle(httpRequest: HttpRequest): HttpResponse = {
+          bodyRequestOnPushServer.set(httpRequest.getBodyAsString)
+          response()
+            .withStatusCode(HttpStatus.SC_CREATED)
+        }
+      })
```

Comment:
I prefer to return a string and keep the mutability with the AtomicReference a (very) local implementation detail.

Something like :

```suggestion
  private def setupPushServerCallback(pushServer: ClientAndServer): String = {
    val bodyRequestOnPushServer: AtomicReference[String] = new AtomicReference()
    // ...
   }
```

Or at the very least we could return the atomic ref:

```suggestion
  private def setupPushServerCallback(pushServer: ClientAndServer): AtomicReference[String] = {
    val bodyRequestOnPushServer: AtomicReference[String] = new AtomicReference()
    // ...
   }
```